examples/cloudsql-custom-metric/main.py

```diff
@@ -0,0 +1,352 @@
+    table_id = create_table_id(project_id, dataset_name, False)
+    table_reference = TableReference.from_string(table_id)
+
+    try:
+        client.get_table(table_reference)
+        return True
+    except NotFound:
+        return False
+
+
+@logged
+def make_bigquery_query(query, location):
+    client = big_query_client()
+
+    query_job = client.query(
+        query=query,
+        location=location,
+    )
+    return query_job
+
+
+@logged
+def process_job_to_timeseries(job):
+
+    cloud_logger.debug("keys in cache before processing: %d",
+                       len(cache.keys()))
+    for row in job:
+        key = row.project_id + "-" + row.database_id
+        cache[key] = row_to_timeseries(row.project_id, row)
+
+    cloud_logger.debug("keys in cache after processing: %d",
+                       len(cache.keys()))
+
+    return [timeseries for timeseries in cache.itervalues()]
+
+
+@logged
+def get_cloudsql_timeseries(bigquery_project,
+                            bigquery_dataset,
+                            all=False,
+                            location="US"):
+    table_id = create_table_id(project_id=bigquery_project,
+                               dataset_name=bigquery_dataset,
+                               all=all)
+    query = QUERY.format(table_id)
+
+    job_result = make_bigquery_query(query=query, location=location)
+    timeseries = process_job_to_timeseries(job_result)
+    return timeseries
+
+
+@logged
+def create_timeseries(project_id, timeseries):
+    try:
+        client = monitoring_client()
+        project_name = client.project_path(project_id)
+        client.create_time_series(project_name, timeseries)
+        cloud_logger.debug("length of timeseries: %d", len(timeseries))
+    except google.api_core.exceptions.InvalidArgument:
+        cloud_logger.warn("unable to create timeseries")
+
+
+@logged
+def main(project_id, dataset, location="US", all=False, forever=False):
+    setup_logging()
+    create_custom_metric(project_id)
+
+    # If not a forever, run once and exit
+    if not forever:
+        cloud_logger.debug("not running forever. all is `{0}`".format(all))
+        timeseries = get_cloudsql_timeseries(
+                                    project_id,
+                                    dataset,
+                                    all,
+                                    location)
+        create_timeseries(project_id, timeseries)
+    # If it is a forever, we'll get all the records once,
+    # then just query todays table every minute
+    else:
+        cloud_logger.debug("running forever")
+        timeseries = get_cloudsql_timeseries(
+                                    project_id,
+                                    dataset,
+                                    True,
+                                    location)
+        create_timeseries(project_id, timeseries)
+
+        ticker = threading.Event()
+        while not ticker.wait(WAIT_TIME_SECONDS):
```

Comment:
Is there a reason to have a long running thread to do this forever? It seems there's an opportunity to use a [Cron Job](https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/). This would prevent one failure from killing subsequent runs and make it clearer what failed when.

Reply:
The challenge I have here is that on the first run, I want the query to be different.  I'll think about this.